core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -556,6 +562,42 @@ public Object readGenericValue() throws IOException {
         return null;
     }
 
+    /**
+     * Reads a {@link NamedWriteable} from the current stream, by first reading its name and then looking for
+     * the corresponding entry in the registry by name, so that the proper object can be read and returned.
+     */
+    <C> C readNamedWriteable(String category) throws IOException {
+        String name = readString();
+        NamedWriteable<C> namedWriteable = namedWriteableRegistry.getPrototype(category, name);
+        return namedWriteable.readFrom(this);
```

Comment:
It's not a big deal, but I liked that it allows to
 - make StreamInput not depend on NamedWriteableRegistry
 - stack registries: you could stack registers on top on each other, while the current approach requires that you provide a registry that knows about all namedwriteables at once. I don't have use-cases in mind, but thought this could be useful.